mjpc/planners/cross_entropy/planner.cc

```diff
@@ -0,0 +1,561 @@
+      {mjITEM_END}};
+
+  // set number of trajectory slider limits
+  mju::sprintf_arr(defCrossEntropy[0].other, "%i %i", 1, kMaxTrajectory);
+
+  // set spline point limits
+  mju::sprintf_arr(defCrossEntropy[2].other, "%i %i", MinSamplingSplinePoints,
+                   MaxSamplingSplinePoints);
+
+  // set noise standard deviation limits
+  mju::sprintf_arr(defCrossEntropy[3].other, "%f %f", MinNoiseStdDev,
+                   MaxNoiseStdDev);
+
+  // add sampling planner
+  mjui_add(&ui, defCrossEntropy);
+}
+
+// planner-specific plots
+void CrossEntropyPlanner::Plots(mjvFigure* fig_planner, mjvFigure* fig_timer,
+                                int planner_shift, int timer_shift,
+                                int planning, int* shift) {
+  // ----- planner ----- //
+  double planner_bounds[2] = {-6.0, 6.0};
+
+  // improvement
+  mjpc::PlotUpdateData(fig_planner, planner_bounds,
+                       fig_planner->linedata[0 + planner_shift][0] + 1,
+                       mju_log10(mju_max(improvement, 1.0e-6)), 100,
+                       0 + planner_shift, 0, 1, -100);
+
+  // legend
+  mju::strcpy_arr(fig_planner->linename[0 + planner_shift], "Improvement");
+
+  fig_planner->range[1][0] = planner_bounds[0];
+  fig_planner->range[1][1] = planner_bounds[1];
+
+  // bounds
+  double timer_bounds[2] = {0.0, 1.0};
+
+  // ----- timer ----- //
+
+  PlotUpdateData(
+      fig_timer, timer_bounds, fig_timer->linedata[0 + timer_shift][0] + 1,
+      1.0e-3 * noise_compute_time * planning, 100, 0 + timer_shift, 0, 1, -100);
+
+  PlotUpdateData(fig_timer, timer_bounds,
+                 fig_timer->linedata[1 + timer_shift][0] + 1,
+                 1.0e-3 * rollouts_compute_time * planning, 100,
+                 1 + timer_shift, 0, 1, -100);
+
+  PlotUpdateData(fig_timer, timer_bounds,
+                 fig_timer->linedata[2 + timer_shift][0] + 1,
+                 1.0e-3 * policy_update_compute_time * planning, 100,
+                 2 + timer_shift, 0, 1, -100);
+
+  // legend
+  mju::strcpy_arr(fig_timer->linename[0 + timer_shift], "Noise");
+  mju::strcpy_arr(fig_timer->linename[1 + timer_shift], "Rollout");
+  mju::strcpy_arr(fig_timer->linename[2 + timer_shift], "Policy Update");
+
+  // planner shift
+  shift[0] += 1;
+
+  // timer shift
+  shift[1] += 3;
+}
+
+double CrossEntropyPlanner::CandidateScore(int candidate) const {
+  return trajectory[trajectory_order[candidate]].total_return;
+}
+
+// set action from candidate policy
+void CrossEntropyPlanner::ActionFromCandidatePolicy(double* action,
+                                                    int candidate,
+                                                    const double* state,
+                                                    double time) {
+  candidate_policy[trajectory_order[candidate]].Action(action, state, time);
+}
+
+void CrossEntropyPlanner::CopyCandidateToPolicy(int candidate) {
+  return;  // unused, included for API compliance
+}
```

Comment:
new: this used to be called at the end of `OptimizePolicy`, which would cause the `policy` to be set equal to the `winner` - this isn't correct behavior.

Reply:
I removed this after moving from RankedPlanner -> Planner.